stan/math/prim/meta/holder.hpp

```diff
@@ -261,17 +262,34 @@ auto holder_handle_element(T& a, T*& res) {
   res = &a;
   return std::make_tuple();
 }
+template <typename T,
+          std::enable_if_t<!(Eigen::internal::traits<std::decay_t<T>>::Flags
+                             & Eigen::NestByRefBit)>* = nullptr>
+auto holder_handle_element(T&& a, std::remove_reference_t<T>*& res) {
+  res = &a;
+  return std::make_tuple();
+}
 
 /**
- * Handles single element (moving rvalues to heap) for construction of
- * `holder` or `holder_cl` from a functor. Rvalue is moved to heap and the
- * pointer to heap memory is assigned to res and returned in a tuple.
+ * Handles single element (moving rvalue non-expressions to heap) for
+ * construction of `holder` or `holder_cl` from a functor. Rvalue non-expression
+ * is moved to heap and the pointer to heap memory is assigned to res and
+ * returned in a tuple.
  * @tparam T type of the element
  * @param a element to handle
  * @param res resulting pointer to element
  * @return tuple of pointers allocated on heap (containing single pointer).
  */
-template <typename T, require_t<std::is_rvalue_reference<T&&>>* = nullptr>
+template <typename T, require_t<std::is_rvalue_reference<T&&>>* = nullptr,
+          std::enable_if_t<
+              static_cast<bool>(Eigen::internal::traits<std::decay_t<T>>::Flags&
+                                    Eigen::NestByRefBit)>* = nullptr>
```

Comment:
What is a `NestByRefBit`? This should probably go in the docs. I didn't find it searching Google

Reply:
Eigen is stuck in C++03, in which rvalue references and perfect forwarding do not exist. So when constructing expressions, they can not know which arguments are rvalues and might go out of scope before the expression is evaled. Instead they decide which types to store by reference and which by value depending on the argument type. Lightweight objects, such as operations do not set `NestByRefBit` and are stored by value. Heavy objects, such as matrices set `NestByRefBit` and are stored by reference.